test/functional/rpc_blockchain.py

```diff
@@ -451,6 +452,7 @@ def _test_getnetworkhashps(self):
         )
         # This should be 2 hashes every 10 minutes or 1/300
         assert abs(hashes_per_second * 300 - 1) < 0.0001
+        assert abs(hashes_per_second_since_diff_adjustment * 300 - 1) < 0.006
```

Comment:
Even though there is an assumption that the blocks in `node[0]` will be 200 in `rpc_blockchain.py`, I am not a fan of the hardcoded `0.006`.

Since you are testing that whenever the first parameter of `getnetworkhashps` is negative, it will return the average hashes per second since the last difficulty change, why not calculate the number of blocks since the last difficulty change, call the `getnetworkhashps` with that, and then check against the `getnetworkhashps` rpc call with a negative value? 
This approach is better than the hardcoded `0.006`. Also, if the assumption of `200` blocks changes, you would not need to update the `0.006` value.
Also in the below diff, I added @pablomartin4btc comment https://github.com/bitcoin/bitcoin/pull/27852/commits/dc190c95fcb716a92b7c4fbcc10f30445a2689fb#r1332416752

<details>
<summary> see diff</summary>

```diff
diff --git a/test/functional/rpc_blockchain.py b/test/functional/rpc_blockchain.py
index 8ae1e661d4..45e5cff8cd 100755
--- a/test/functional/rpc_blockchain.py
+++ b/test/functional/rpc_blockchain.py
@@ -58,6 +58,7 @@ TIME_RANGE_STEP = 600  # ten-minute steps
 TIME_RANGE_MTP = TIME_GENESIS_BLOCK + (HEIGHT - 6) * TIME_RANGE_STEP
 TIME_RANGE_TIP = TIME_GENESIS_BLOCK + (HEIGHT - 1) * TIME_RANGE_STEP
 TIME_RANGE_END = TIME_GENESIS_BLOCK + HEIGHT * TIME_RANGE_STEP
+DIFFICULTY_ADJUSTMENT_INTERVAL = 2016
 
 
 class BlockchainTest(BitcoinTestFramework):
@@ -437,7 +438,6 @@ class BlockchainTest(BitcoinTestFramework):
     def _test_getnetworkhashps(self):
         self.log.info("Test getnetworkhashps")
         hashes_per_second = self.nodes[0].getnetworkhashps()
-        hashes_per_second_since_diff_adjustment = self.nodes[0].getnetworkhashps(-1, -1)
         assert_raises_rpc_error(
             -3,
             textwrap.dedent("""
@@ -451,7 +451,15 @@ class BlockchainTest(BitcoinTestFramework):
         )
         # This should be 2 hashes every 10 minutes or 1/300
         assert abs(hashes_per_second * 300 - 1) < 0.0001
-        assert abs(hashes_per_second_since_diff_adjustment * 300 - 1) < 0.006
+
+        # Test setting the first param of getnetworkhashps to negative value returns the average network
+        # hashes per second from the last difficulty change.
+        current_block_height = self.nodes[0].getmininginfo()['blocks']
+        blocks_since_last_diff_change = current_block_height % DIFFICULTY_ADJUSTMENT_INTERVAL + 1
+        expected_hashes_per_second_since_diff_change = self.nodes[0].getnetworkhashps(blocks_since_last_diff_change)
+
+        assert_equal(self.nodes[0].getnetworkhashps(-1), expected_hashes_per_second_since_diff_change)
+        assert_equal(self.nodes[0].getnetworkhashps(-2), expected_hashes_per_second_since_diff_change)
 
     def _test_stopatheight(self):
         self.log.info("Test stopping at height")
```
</details>

feel free to cherry-pick https://github.com/bitcoin/bitcoin/commit/17e505abfe1557a8e5c1819c9d09d27abece25b9

Reply:
I do agree with this approach, avoids the hassle of updating the hardcoded value if it ever changes and adds more clarity of the context (blockchain setup in the test).

Reply:
updated and used the cherry-picked commit 17e505a thanks for the reviews!